File: android/src/main/java/com/admost/reactnative/AdMostInterstitialModule.java
package com.admost.reactnative;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;

import java.util.HashMap;

import admost.sdk.AdMostInterstitial;
import admost.sdk.listener.AdMostAdListener;

public class AdMostInterstitialModule extends ReactContextBaseJavaModule {

  private static final String TAG = AdMostInterstitialModule.class.getSimpleName();
  private final ReactApplicationContext reactContext;
  private final HashMap<String, AdMostInterstitial> adMostInterstitialMap; // zoneId -> AdMostInterstitial

  public AdMostInterstitialModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
    this.adMostInterstitialMap = new HashMap<>();
  }

  @NonNull
  @Override
  public String getName() {
    return "AdMostInterstitial";
  }

  @ReactMethod
  public void loadAd(final String zoneId) {
    AdMostInterstitial adMostInterstitial = adMostInterstitialMap.get(zoneId);

    if (adMostInterstitial == null) {
      AdMostInterstitial newAdMostInterstitial = new AdMostInterstitial(getCurrentActivity(), zoneId, getListener(zoneId));
      adMostInterstitialMap.put(zoneId, newAdMostInterstitial);
      newAdMostInterstitial.refreshAd(false);
      return;
    }

    adMostInterstitial.refreshAd(false);
  }

  @ReactMethod
  public void destroyAd(final String zoneId) {
    AdMostInterstitial adMostInterstitial = adMostInterstitialMap.get(zoneId);

    if (adMostInterstitial != null) {
      adMostInterstitial.destroy();
      adMostInterstitialMap.remove(zoneId);
    }
  }

  @ReactMethod
  public void showAd(final String zoneId, final Promise promise) {
    AdMostInterstitial adMostInterstitial = adMostInterstitialMap.get(zoneId);

    if (adMostInterstitial != null) {
      adMostInterstitial.show();
      promise.resolve(true);
    } else {
      promise.reject(Constants.ADMOST_INSTANCE_NOT_FOUND, "Couldn't find any instance in this zone, call loadAd");
    }
  }

  @ReactMethod
  public void isLoading(final String zoneId, final Promise promise) {
    AdMostInterstitial adMostInterstitial = adMostInterstitialMap.get(zoneId);

    if (adMostInterstitial != null) {
      promise.resolve(adMostInterstitial.isLoading());
    } else {
      promise.reject(Constants.ADMOST_INSTANCE_NOT_FOUND, "Couldn't find any instance in this zone, call loadAd");
    }
  }

  @ReactMethod
  public void isLoaded(final String zoneId, final Promise promise) {
    AdMostInterstitial adMostInterstitial = adMostInterstitialMap.get(zoneId);

    if (adMostInterstitial != null) {
      promise.resolve(adMostInterstitial.isLoaded());
    } else {
      promise.reject(Constants.ADMOST_INSTANCE_NOT_FOUND, "Couldn't find any instance in this zone, call loadAd");
    }
  }

  private AdMostAdListener getListener(final String zoneId) {
    return new AdMostAdListener() {
      @Override
      public void onReady(String network, int ecpm) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putInt("ecpm", ecpm);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_INTERSTITIAL_ON_READY, params);
      }

      @Override
      public void onFail(int errorCode) {
        WritableMap params = Arguments.createMap();
        params.putInt("errorCode", errorCode);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_INTERSTITIAL_ON_FAIL, params);
      }

      @Override
      public void onDismiss(String message) {
        WritableMap params = Arguments.createMap();
        params.putString("message", message);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_INTERSTITIAL_ON_DISMISS, params);
      }

      @Override
      public void onComplete(String network) {

      }

      @Override
      public void onShown(String network) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_INTERSTITIAL_ON_SHOWN, params);
      }

      @Override
      public void onClicked(String network) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_INTERSTITIAL_ON_CLICKED, params);
      }

      @Override
      public void onStatusChanged(int statusCode) {
        WritableMap params = Arguments.createMap();
        params.putInt("statusCode", statusCode);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_INTERSTITIAL_ON_STATUS_CHANGED, params);
      }
    };
  }

}

File: modules/react-native-admost/ios/AdMostInterstitialModule.swift
import AMRSDK
import AppTrackingTransparency
import AdSupport

@objc(AdMostInterstitialModule)
class AdMostInterstitialModule: NSObject, AMRInterstitialDelegate {

    var adMostInterstitialDict: Dictionary<String, AMRInterstitial> = [:] // zoneId -> AMRInterstitial

    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc
    func loadAd(_ zoneId: NSString) -> Void {
        let adMostInterstitial = adMostInterstitialDict[zoneId as String]

        if adMostInterstitial == nil {
            let newAdMostInterstitial: AMRInterstitial = AMRInterstitial(forZoneId: zoneId as String)
            newAdMostInterstitial.delegate = self
            adMostInterstitialDict[zoneId as String] = newAdMostInterstitial

            newAdMostInterstitial.load()
            return;
        }

        adMostInterstitial?.load()
    }

    @objc
    func destroyAd(_ zoneId: NSString) -> Void {
        let adMostInterstitial = adMostInterstitialDict[zoneId as String]

        if adMostInterstitial != nil {
            // todo: destroy adMostAd
            adMostInterstitialDict.removeValue(forKey: zoneId as String)
        }
    }

    @objc
    func showAd(
            _ zoneId: NSString,
            tag adTag: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostInterstitial = adMostInterstitialDict[zoneId as String]

        if adMostInterstitial != nil {
            adMostInterstitial?.show(from: UIApplication.shared.delegate?.window??.rootViewController, withTag: adTag as String)
            resolve(true)
        } else {
            reject("ADMOST_INSTANCE_NOT_FOUND", "Couldn't find any instance in this zone, call loadAd", nil);
        }
    }

    @objc
    func isLoading(
            _ zoneId: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostInterstitial = adMostInterstitialDict[zoneId as String]

        if adMostInterstitial != nil {
            resolve(adMostInterstitial?.isLoading)
        } else {
            reject("ADMOST_INSTANCE_NOT_FOUND", "Couldn't find any instance in this zone, call loadAd", nil);
        }
    }

    @objc
    func isLoaded(
            _ zoneId: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostInterstitial = adMostInterstitialDict[zoneId as String]

        if adMostInterstitial != nil {
            resolve(adMostInterstitial?.isLoaded)
        } else {
            reject("ADMOST_INSTANCE_NOT_FOUND", "Couldn't find any instance in this zone, call loadAd", nil);
        }
    }

    func didReceive(_ interstitial: AMRInterstitial!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_INTERSTITIAL_ON_READY",
                body: ["network": interstitial.networkName!, "ecpm": interstitial.ecpm!, "zoneId": interstitial.zoneId!]
        )
    }

    func didFail(toReceive interstitial: AMRInterstitial!, error: AMRError!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_INTERSTITIAL_ON_FAIL",
                body: ["errorDescription": error.errorDescription!, "errorCode": error.errorCode, "zoneId": interstitial.zoneId!]
        )
    }
    
    func didDismiss(_ interstitial: AMRInterstitial!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_INTERSTITIAL_ON_DISMISS",
                body: ["zoneId": interstitial.zoneId!]
        )
    }

    func didShow(_ interstitial: AMRInterstitial!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_INTERSTITIAL_ON_SHOWN",
                body: ["network": interstitial.networkName!, "zoneId": interstitial.zoneId!]
        )
    }

    func didClick(_ interstitial: AMRInterstitial!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_INTERSTITIAL_ON_CLICKED",
                body: ["network": interstitial.networkName!, "zoneId": interstitial.zoneId!]
        )
    }

    func didInterstitialStateChanged(_ interstitial: AMRInterstitial!, state: AMRAdState) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_INTERSTITIAL_ON_STATUS_CHANGED",
                body: ["zoneId": interstitial.zoneId!, "statusCode": state]
        )
    }
}
